### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by xla**
This change-set tries to introduce consistent way of logging and makes it a concern which can be mixed in, rather than being interleaved with other logic. it does so by implementing a handler and an interface implementation for DNS and the store respectively. Basic assumption was that we want high-level transaction logging for the edge (for the interface glimpse will offer like DNS now and potentially HTTP in the future) and logging of errors on the component level (e.g. store).

Looking for feedback around the log format which currently is space separated strings.
***
@peterbourgon @grobie

### CODE REVIEW

File: agent/logging.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+			start:          time.Now(),
+		}, r)
+	}
+}
+
+type loggingWriter struct {
+	dns.ResponseWriter
+
+	logger *log.Logger
+	start  time.Time
```
Comment (peterbourgon) [inline agent/logging.go]: I would take this out of the struct, to keep the struct absolutely minimal, and set it as the first expression in `WriteMsg`.

File: agent/logging.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+	return err
+}
+
+type loggingStore struct {
+	logger *log.Logger
+	next   store
+}
+
+func newLoggingStore(logger *log.Logger, next store) *loggingStore {
```
Comment (peterbourgon) [inline agent/logging.go]: This should return type store. (Hint: update `golint` and run it on this file.)

File: agent/logging.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+func (s *loggingStore) getServers(zone string) (is instances, err error) {
+	defer func(start time.Time) {
+		s.log(start, "getServers", err, zone)
+	}(time.Now())
+
+	return s.next.getServers(zone)
+}
+
+func (s *loggingStore) log(start time.Time, op string, err error, input string) {
```
Comment (peterbourgon) [inline agent/logging.go]: This should better take as a param `took time.Duration`...

File: agent/logging.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+func newLoggingStore(logger *log.Logger, next store) *loggingStore {
+	return &loggingStore{
+		logger: logger,
+		next:   next,
+	}
+}
+
+func (s *loggingStore) getInstances(i info) (is instances, err error) {
+	defer func(start time.Time) {
+		s.log(start, "getInstances", err, i.addr())
```
Comment (peterbourgon) [inline agent/logging.go]: ...and re: below, `s.log(time.Since(start), ...)`

File: agent/logging.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+func (s *loggingStore) log(start time.Time, op string, err error, input string) {
+	if err == nil {
+		return
+	}
+
+	label := errToLabel[errUntracked]
+
+	switch e := err.(type) {
+	case *glimpseError:
```
Comment (peterbourgon) [inline agent/logging.go]: err would be a pointer-to-glimpseError?

File: agent/types.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 var (
 	errConsulAPI   = errors.New("Consul API failed")
 	errInvalidIP   = errors.New("invalid IP address")
-	errNoInstances = errors.New("no instances found")
+	errNoInstances = errors.New("no instances")
```
Comment (peterbourgon) [inline agent/types.go]: :+1:

**Comment by peterbourgon** [comment]
In general I like the approach. :dango: once those minor things are addressed.

File: agent/logging.go (hunk 1 above)
Comment (xla) [inline agent/logging.go]: That would create a different result. `WriteMsg` could be called way later in the life of a request. In its current form we make sure it's at the beginning of the request.

File: agent/logging.go (hunk 5 above)
Comment (xla) [inline agent/logging.go]: Yes, origin [here](https://github.com/soundcloud/glimpse/blob/master/agent/types.go#L118).

File: agent/logging.go (hunk 1 above)
Comment (peterbourgon) [inline agent/logging.go]: Ah! Full ack.

File: agent/logging.go (hunk 5 above)
Comment (peterbourgon) [inline agent/logging.go]: Why reference and not value?

Comment (peterbourgon) [inline agent/logging.go]: [Context/explanation](http://play.golang.org/p/2NgMTRu7fU).

Comment (xla) [inline agent/logging.go]: Good feedback, will adjust outside of this PR.

**Pull request merged**